File: Offwind.WakeFarmControl/FarmControlRasmus/Compute_Vell.m
% Wake Simulation
% (C) Rasmus Christensen
% Control and Automation, Aalborg University 2014

% Compute the velocity in front of each wind-turbine, with respect to the wind input.

function [vel_output] = Compute_Vell(yTurb ,xTurbC ,yTurbC ,x ,wField ,Uhub ,kWake ,iMax ,jMax ,nTurb ,dTurb ,Ct, dy)
    vell_i = wField;
    
    shadow = zeros(1,nTurb);

	r0 = 0.5*dTurb;
	nk = 2*ceil(dTurb/dy);

    for k = 1:nTurb
		J = 0;
		SS = 0;
		SS0 = pi*r0*r0;

        for i = 1:k-1
			RR_i = r0 + kWake*(x(xTurbC(k)) - x(xTurbC(i)));
			Dij = abs(yTurb(i)-yTurb(k));

			if (RR_i >= (r0 + Dij)) || (Dij <= dy)
				SS = SS+((r0*r0)/(RR_i*RR_i));
			else
				if RR_i >= (r0 + Dij) && Dij > dy
					J 			= J+1;
					Alpha_i 	= acos((RR_i*RR_i)+(Dij*Dij)-(r0*r0)/(2*RR_i*Dij));
					Alpha_k 	= acos(((r0*r0)+(Dij*Dij)-(RR_i*RR_i))/(2*r0*Dij));
					Area 		= AArea(RR_i, r0, Dij);
					shadow(J) 	= (Alpha_i*(RR_i*RR_i) + Alpha_k*(r0*r0))- 2*Area;
					SS 			= SS + ((shadow(J))/SS0)*((r0*r0)/(RR_i*RR_i));
				end
			end
        end

        for ii = xTurbC(k):iMax
			rrt 	= r0 + kWake*(x(ii)-x(xTurbC(k)));
			nj 		= ceil(rrt/dy);

			jjMin   = floor(max(1, yTurbC(k)-nj));
			jjMax   = ceil(min([jMax, yTurbC(k)+nj]));

            for j = jjMin:jjMax
                if ((-vell_i(ii,j)+Uhub(k)) > 0) && (ii > xTurbC(k)+nk)
					vell_i(ii,j) = min([vell_i(ii-1,j), Uhub(k) + Uhub(k)*(sqrt(1-Ct(k))-1)*((r0*r0)/(rrt*rrt))*(1-(1-sqrt(1-Ct(k)))*SS)]);
                    vell_i(ii,j) = max([0, vell_i(ii,j)]);
				else
					vell_i(ii,j) = (Uhub(k) + Uhub(k)*(sqrt(1-Ct(k))-1)*(r0/rrt)*(r0/rrt))*(1-(1-sqrt(1-Ct(k)))*SS);
                    vell_i(ii,j) = max([0, vell_i(ii,j)]);
                end
            end
        end
    end
    
	vel_output = vell_i;

end

function [area_out] = AArea(x, y, z) % Internal function to compute the shadowed area.
	PP = (x+y+z)*0.5;
	area_out = sqrt(PP*(PP-x)*(PP-y)*(PP-z));
end

File: Offwind.WakeFarmControl/WakeFarmControlRasmus/Compute_Vell.cs
using System;
using ILNumerics;
using MatlabInterpreter;

namespace WakeFarmControlR
{
    internal partial class TranslatedCode
    {
        #region "Original function comments"
        // Wake Simulation
        // (C) Rasmus Christensen
        // Control and Automation, Aalborg University 2014

        // Compute the velocity in front of each wind-turbine, with respect to the wind input.
        #endregion
        internal static void Compute_Vell(out ILArray<double> vel_output, ILArray<double> yTurb, ILArray<int> xTurbC, ILArray<int> yTurbC, ILArray<double> x, ILArray<double> wField, ILArray<double> Uhub, double kWake, int iMax, int jMax, int nTurb, double dTurb, ILArray<double> Ct, double dy)
        {
            #region "Used variables declaration"
            ILArray<double> vell_i;
            ILArray<double> shadow;
            double r0;
            double nk;
            int k;
            int J;
            double SS;
            double SS0;
            int i;
            double RR_i;
            double Dij;
            double Alpha_i;
            double Alpha_k;
            double Area;
            int ii;
            double rrt;
            double nj;
            int jjMin;
            int jjMax;
            #endregion


            vell_i = wField.C;

            shadow = zeros(1, nTurb);

            r0 = 0.5 * dTurb;
            nk = 2 * ceil(dTurb / dy);

            for (k = 1; k <= nTurb; k++)
            {
                J = 0;
                SS = 0;
                SS0 = pi * r0 * r0;

                for (i = 1; i <= k - 1; i++)
                {
                    RR_i = r0 + kWake * (x._(xTurbC._(k)) - x._(xTurbC._(i)));
                    Dij = abs(yTurb._(i) - yTurb._(k));

                    if ((RR_i >= (r0 + Dij)) || (Dij <= dy))
                    {
                        SS = SS + ((r0 * r0) / (RR_i * RR_i));
                    }
                    else
                    {
                        if (RR_i >= (r0 + Dij) && Dij > dy)
                        {
                            J               = J + 1;
                            Alpha_i         = acos((RR_i * RR_i) + (Dij * Dij) - (r0 * r0) / (2 * RR_i * Dij));
                            Alpha_k         = acos(((r0 * r0) + (Dij * Dij) - (RR_i * RR_i)) / (2 * r0 * Dij));
                            AArea(out Area, RR_i, r0, Dij);
                            shadow._(J,     '=', (Alpha_i * (RR_i * RR_i) + Alpha_k * (r0 * r0)) - 2 * Area);
                            SS              = SS + ((shadow._(J)) / SS0) * ((r0 * r0) / (RR_i * RR_i));
                        }
                    }
                }

                for (ii = xTurbC._(k); ii <= iMax; ii++)
                {
                    rrt     = r0 + kWake * (x._(ii) - x._(xTurbC._(k)));
                    nj      = ceil(rrt / dy);

                    jjMin   = floor_(max(1, yTurbC._(k) - nj));
                    jjMax   = ceil_(min_(__[ jMax, yTurbC._(k) + nj ]));

                    for (var j = jjMin; j <= jjMax; j++)
                    {
                        if (((-vell_i._(ii, j) + Uhub._(k)) > 0) && (ii > xTurbC._(k) + nk))
                        {
                            vell_i._(ii, j, '=', min_(__[ vell_i._(ii - 1, j), Uhub._(k) + Uhub._(k) * (sqrt(1 - Ct._(k)) - 1) * ((r0 * r0) / (rrt * rrt)) * (1 - (1 - sqrt(1 - Ct._(k))) * SS) ]));
                            vell_i._(ii, j, '=', max_(__[ 0, vell_i._(ii, j) ]));
                        }
                        else
                        {
                            vell_i._(ii, j, '=', (Uhub._(k) + Uhub._(k) * (sqrt(1 - Ct._(k)) - 1) * (r0 / rrt) * (r0 / rrt)) * (1 - (1 - sqrt(1 - Ct._(k))) * SS));
                            vell_i._(ii, j, '=', max_(__[ 0, vell_i._(ii, j) ]));
                        }
                    }
                }
            }

            vel_output = vell_i;
        }

        public static void AArea(out double area_out, double x, double y, double z) // Internal function to compute the shadowed area.
        {
            #region "Used variables declaration"
            double PP;
            #endregion

            PP = (x + y + z) * 0.5;
            area_out = sqrt(PP * (PP - x) * (PP - y) * (PP - z));
        }
    }
}
